File: logica/Prestamos.py
"""
Módulo de préstamos y reservas.

Se mantiene separado de Admin.py a propósito: Admin.py concentra el
catálogo (libros + estantería) y usuarios del sistema (login). Este
archivo concentra el movimiento de la biblioteca (quién tiene qué libro
y quién lo está esperando), que es un dominio distinto.

Estructura de cada préstamo en prestamos.json:
    {
        "id_libro": "L001",
        "usuario": "Juan Pérez",        # nombre libre del lector, NO es
                                         # una cuenta de usuarios.json
        "fecha_prestamo": "2026-06-10",
        "fecha_limite": "2026-06-17",
        "fecha_devolucion": null,       # se llena al devolver
        "estado": "activo"              # activo | devuelto | vencido
    }

Estructura de cada reserva en reservas.json:
    {
        "id_libro": "L001",
        "usuario": "María G.",
        "fecha_solicitud": "2026-06-15",
        "estado": "pendiente"           # pendiente | aprobada | rechazada
    }
"""

import json
import os
from datetime import datetime, timedelta

CARPETA_RAIZ = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
CARPETA_DATOS = os.path.join(CARPETA_RAIZ, "datos")
ARCHIVO_PRESTAMOS = os.path.join(CARPETA_DATOS, "prestamos.json")
ARCHIVO_RESERVAS = os.path.join(CARPETA_DATOS, "reservas.json")

DIAS_PRESTAMO_DEFAULT = 7  # plazo estándar al registrar un préstamo nuevo

lista_prestamos = []
lista_reservas = []


# =========================================================================
# CARGA / GUARDADO — mismo patrón que cargar_desde_archivo() en Admin.py
# =========================================================================
def cargar_prestamos():
    global lista_prestamos
    lista_prestamos = []
    os.makedirs(CARPETA_DATOS, exist_ok=True)

    if not os.path.exists(ARCHIVO_PRESTAMOS):
        return

    try:
        with open(ARCHIVO_PRESTAMOS, "r", encoding="utf-8") as f:
            lista_prestamos = json.load(f)
    except Exception as e:
        print(f"Aviso al cargar préstamos: {e}")

    _actualizar_estados_vencidos()


def guardar_prestamos():
    try:
        os.makedirs(CARPETA_DATOS, exist_ok=True)
        with open(ARCHIVO_PRESTAMOS, "w", encoding="utf-8") as f:
            json.dump(lista_prestamos, f, ensure_ascii=False, indent=2)
    except Exception as e:
        print(f"Error al guardar préstamos: {e}")


def cargar_reservas():
    global lista_reservas
    lista_reservas = []
    os.makedirs(CARPETA_DATOS, exist_ok=True)

    if not os.path.exists(ARCHIVO_RESERVAS):
        return

    try:
        with open(ARCHIVO_RESERVAS, "r", encoding="utf-8") as f:
            lista_reservas = json.load(f)
    except Exception as e:
        print(f"Aviso al cargar reservas: {e}")


def guardar_reservas():
    try:
        os.makedirs(CARPETA_DATOS, exist_ok=True)
        with open(ARCHIVO_RESERVAS, "w", encoding="utf-8") as f:
            json.dump(lista_reservas, f, ensure_ascii=False, indent=2)
    except Exception as e:
        print(f"Error al guardar reservas: {e}")


# =========================================================================
# LÓGICA DE ESTADOS
# =========================================================================
def _hoy_str():
    return datetime.now().strftime("%Y-%m-%d")


def _actualizar_estados_vencidos():
    """Recalcula 'vencido' para préstamos activos cuya fecha_limite ya pasó.
    Se llama automáticamente al cargar, así el estado siempre refleja la
    fecha real del sistema sin depender de que alguien lo marque a mano."""
    hoy = datetime.now().date()
    cambio = False
    for p in lista_prestamos:
        if p["estado"] == "activo":
            limite = datetime.strptime(p["fecha_limite"], "%Y-%m-%d").date()
            if hoy > limite:
                p["estado"] = "vencido"
                cambio = True
    if cambio:
        guardar_prestamos()


def registrar_prestamo(id_libro, usuario, dias=DIAS_PRESTAMO_DEFAULT):
    """Crea un nuevo préstamo activo. No valida aquí si el libro ya está
    prestado: esa validación de negocio vive en la capa de UI, que tiene
    acceso a lista_libros para cruzar datos."""
    hoy = datetime.now()
    limite = hoy + timedelta(days=dias)
    nuevo = {
        "id_libro": id_libro,
        "usuario": usuario,
        "fecha_prestamo": hoy.strftime("%Y-%m-%d"),
        "fecha_limite": limite.strftime("%Y-%m-%d"),
        "fecha_devolucion": None,
        "estado": "activo",
    }
    lista_prestamos.append(nuevo)
    guardar_prestamos()
    return nuevo


def registrar_devolucion(id_libro):
    """Marca como devuelto el préstamo activo/vencido más reciente de ese
    libro. Retorna el registro actualizado, o None si no había préstamo
    abierto para ese id_libro."""
    candidatos = [p for p in lista_prestamos if p["id_libro"] == id_libro and p["estado"] in ("activo", "vencido")]
    if not candidatos:
        return None
    prestamo = candidatos[-1]
    prestamo["fecha_devolucion"] = _hoy_str()
    prestamo["estado"] = "devuelto"
    guardar_prestamos()
    return prestamo


def prestamo_activo_de(id_libro):
    """Devuelve el préstamo activo/vencido de un libro, si existe."""
    for p in lista_prestamos:
        if p["id_libro"] == id_libro and p["estado"] in ("activo", "vencido"):
            return p
    return None


def contar_vencidos():
    return sum(1 for p in lista_prestamos if p["estado"] == "vencido")


def contar_activos():
    return sum(1 for p in lista_prestamos if p["estado"] in ("activo", "vencido"))


def actividad_reciente(limite=5):
    """Últimos movimientos (préstamos y devoluciones) ordenados por fecha
    descendente, para el widget de 'Actividad reciente' del dashboard."""
    eventos = []
    for p in lista_prestamos:
        eventos.append({
            "tipo": "prestamo",
            "fecha": p["fecha_prestamo"],
            "usuario": p["usuario"],
            "id_libro": p["id_libro"],
        })
        if p["fecha_devolucion"]:
            eventos.append({
                "tipo": "devolucion",
                "fecha": p["fecha_devolucion"],
                "usuario": p["usuario"],
                "id_libro": p["id_libro"],
            })
    eventos.sort(key=lambda e: e["fecha"], reverse=True)
    return eventos[:limite]


# =========================================================================
# RESERVAS
# =========================================================================
def registrar_reserva(id_libro, usuario):
    nueva = {
        "id_libro": id_libro,
        "usuario": usuario,
        "fecha_solicitud": _hoy_str(),
        "estado": "pendiente",
    }
    lista_reservas.append(nueva)
    guardar_reservas()
    return nueva


def contar_pendientes():
    return sum(1 for r in lista_reservas if r["estado"] == "pendiente")


def aprobar_reserva(indice, dias=DIAS_PRESTAMO_DEFAULT):
    """Aprueba una reserva y la convierte automáticamente en préstamo."""
    if not (0 <= indice < len(lista_reservas)):
        return None
    reserva = lista_reservas[indice]
    reserva["estado"] = "aprobada"
    guardar_reservas()
    return registrar_prestamo(reserva["id_libro"], reserva["usuario"], dias)


def rechazar_reserva(indice):
    if not (0 <= indice < len(lista_reservas)):
        return None
    lista_reservas[indice]["estado"] = "rechazada"
    guardar_reservas()
    return lista_reservas[indice]


# =========================================================================
# CONSULTAS POR USUARIO — usadas por el panel de estudiante/invitado
# =========================================================================
def prestamos_de_usuario(nombre_usuario):
    """Devuelve la lista de préstamos activos o vencidos de un usuario."""
    return [p for p in lista_prestamos
            if p["usuario"] == nombre_usuario and p["estado"] in ("activo", "vencido")]


def contar_prestamos_activos_usuario(nombre_usuario):
    """Cantidad de libros que el usuario tiene actualmente en su poder."""
    return len(prestamos_de_usuario(nombre_usuario))


def historial_usuario(nombre_usuario):
    """Todos los préstamos del usuario (activos, vencidos y devueltos)."""
    return [p for p in lista_prestamos if p["usuario"] == nombre_usuario]

File: web/js/prestamos.js
'use strict';
// ============================================================
// prestamos.js — Préstamos, devoluciones, reservas y penalizaciones
// Equivale a: logica/prestamos.py + partes de menu_bibliotecario.py
// ============================================================

const DIAS_PRESTAMO = 7;

/** Devuelve la fecha de hoy en formato YYYY-MM-DD */
const _hoy = () => new Date().toISOString().split('T')[0];

/** Suma días a una fecha string YYYY-MM-DD */
const _sumarDias = (fechaStr, dias) => {
  const d = new Date(fechaStr + 'T12:00:00');
  d.setDate(d.getDate() + dias);
  return d.toISOString().split('T')[0];
};

/** Diferencia en días entre hoy y una fecha string */
const _diasRestantes = (fechaStr) => {
  const fin = new Date(fechaStr + 'T12:00:00');
  const hoy = new Date();
  hoy.setHours(12, 0, 0, 0);
  return Math.ceil((fin - hoy) / (1000 * 60 * 60 * 24));
};

const Prestamos = {
  // =========================================================
  // PRÉSTAMOS
  // =========================================================

  /** Actualiza estados 'activo' → 'vencido' si ya pasó la fecha límite */
  actualizarVencidos() {
    const prestamos = DB.getPrestamos();
    const hoy = _hoy();
    let cambio = false;
    prestamos.forEach(p => {
      if (p.estado === 'activo' && p.fecha_limite < hoy) {
        p.estado = 'vencido';
        cambio = true;
      }
    });
    if (cambio) DB.setPrestamos(prestamos);
  },

  getAll() {
    this.actualizarVencidos();
    return DB.getPrestamos();
  },

  /**
   * Registra un nuevo préstamo activo.
   * Retorna el registro creado.
   */
  /**
   * login: clave única del usuario (session.login). Opcional para préstamos manuales.
   */
  registrar(id_libro, usuario, dias = DIAS_PRESTAMO, login = null) {
    const fecha_prestamo = _hoy();
    const fecha_limite   = _sumarDias(fecha_prestamo, dias);
    const nuevo = {
      id: 'P' + Date.now(),
      id_libro, usuario,
      login: login || usuario,   // login único para filtrar préstamos del usuario
      fecha_prestamo, fecha_limite,
      fecha_devolucion: null,
      estado: 'activo'
    };
    const p = DB.getPrestamos();
    p.push(nuevo);
    DB.setPrestamos(p);
    return nuevo;
  },

  /**
   * Marca como devuelto el préstamo activo/vencido más reciente de ese libro.
   * Retorna el registro o null si no existía.
   */
  devolver(id_libro) {
    const prestamos = DB.getPrestamos();
    const candidatos = prestamos.filter(
      p => p.id_libro === id_libro && (p.estado === 'activo' || p.estado === 'vencido')
    );
    if (!candidatos.length) return null;
    const prestamo = candidatos[candidatos.length - 1];
    prestamo.fecha_devolucion = _hoy();
    prestamo.estado = 'devuelto';
    DB.setPrestamos(prestamos);
    return prestamo;
  },

  /** Retorna el préstamo activo/vencido de un libro, o null */
  prestamoActivo(id_libro) {
    return this.getAll().find(
      p => p.id_libro === id_libro && (p.estado === 'activo' || p.estado === 'vencido')
    ) || null;
  },

  contarVencidos() { return this.getAll().filter(p => p.estado === 'vencido').length; },
  contarActivos()  { return this.getAll().filter(p => p.estado === 'activo' || p.estado === 'vencido').length; },

  /** Últimos movimientos ordenados por fecha desc (para actividad reciente) */
  actividadReciente(limite = 6) {
    const eventos = [];
    this.getAll().forEach(p => {
      eventos.push({ tipo: 'prestamo', fecha: p.fecha_prestamo, usuario: p.usuario, id_libro: p.id_libro });
      if (p.fecha_devolucion) {
        eventos.push({ tipo: 'devolucion', fecha: p.fecha_devolucion, usuario: p.usuario, id_libro: p.id_libro });
      }
    });
    eventos.sort((a, b) => b.fecha.localeCompare(a.fecha));
    return eventos.slice(0, limite);
  },

  /** Préstamos activos/vencidos de un usuario específico (filtra por login o por nombre para compatibilidad) */
  deUsuario(login) {
    return this.getAll().filter(
      p => (p.login === login || p.usuario === login) && (p.estado === 'activo' || p.estado === 'vencido')
    );
  },

  /** Historial completo (todos los estados) de un usuario */
  historialUsuario(login) {
    return this.getAll().filter(p => p.login === login || p.usuario === login);
  },

  // =========================================================
  // RESERVAS
  // =========================================================

  getReservas() { return DB.getReservas(); },

  registrarReserva(id_libro, usuario) {
    const nueva = {
      id: 'R' + Date.now(),
      id_libro, usuario,
      fecha_solicitud: _hoy(),
      estado: 'pendiente'
    };
    const r = DB.getReservas();
    r.push(nueva);
    DB.setReservas(r);
    return nueva;
  },

  contarPendientes() { return DB.getReservas().filter(r => r.estado === 'pendiente').length; },

  /** Aprueba una reserva y genera el préstamo correspondiente */
  aprobarReserva(id, dias = DIAS_PRESTAMO) {
    const reservas = DB.getReservas();
    const r = reservas.find(x => x.id === id);
    if (!r || r.estado !== 'pendiente') return null;
    r.estado = 'aprobada';
    DB.setReservas(reservas);
    return this.registrar(r.id_libro, r.usuario, dias);
  },

  rechazarReserva(id) {
    const reservas = DB.getReservas();
    const r = reservas.find(x => x.id === id);
    if (!r) return null;
    r.estado = 'rechazada';
    DB.setReservas(reservas);
    return r;
  },

  // =========================================================
  // PENALIZACIONES
  // Equivale a: _estado_penalizacion en menu_usuario.py
  // =========================================================

  estadoPenalizacion(login) {
    const usuario = DB.getUsuario(login);
    if (!usuario) return { puede: true, motivo: '✅ Puedes solicitar libros normalmente.' };

    const strikes        = usuario.strikes || 0;
    const deuda          = usuario.deuda || 0;
    const suspHasta      = usuario.suspendido_hasta || null;

    if (strikes >= 4) {
      return { puede: false, motivo: `🚫 Bloqueo permanente: tienes ${strikes} strikes. Contacta al administrador.` };
    }
    if (suspHasta) {
      const dias = _diasRestantes(suspHasta);
      if (dias > 0) {
        return { puede: false, motivo: `⏳ Suspendido hasta ${suspHasta} (${dias} día(s) restantes). Motivo: 3 strikes acumulados.` };
      }
    }
    if (deuda > 0) {
      return { puede: false, motivo: `💰 Tienes una deuda de $${deuda.toFixed(2)}. Paga en ventanilla para reactivar tus préstamos.` };
    }
    return { puede: true, motivo: '✅ Puedes solicitar libros normalmente.' };
  },

  aplicarStrike(login) {
    const usuario = DB.getUsuario(login);
    if (!usuario) return { ok: false, msg: `El usuario '${login}' no existe.` };

    usuario.strikes = (usuario.strikes || 0) + 1;

    // Al acumular 3 strikes → suspensión de 7 días
    if (usuario.strikes === 3) {
      usuario.suspendido_hasta = _sumarDias(_hoy(), 7);
    }
    DB.saveUsuario(login, usuario);
    return { ok: true, strikes: usuario.strikes };
  },

  limpiarPenalizacion(login) {
    const usuario = DB.getUsuario(login);
    if (!usuario) return { ok: false };
    if (usuario.strikes > 0) usuario.strikes = Math.max(0, usuario.strikes - 1);
    usuario.suspendido_hasta = null;
    usuario.deuda = 0;
    DB.saveUsuario(login, usuario);
    return { ok: true };
  }
};
